improver/utilities/flatten.py

```diff
@@ -4,27 +4,27 @@
 # See LICENSE in the root of the repository for full licensing details.
 """ Provides support utilities for flattening."""
 
-from typing import List, Tuple, Union
+from typing import Any, Generator, Iterable, Union
 
 
-def flatten(nested_list: Union[List, Tuple]) -> List:
-    """Flatten an arbitrarily nested iterable.
+def flatten(nested_iterable: Union[Iterable[Any], Any]) -> Generator[Any, None, None]:
+    """
+    Flatten an arbitrarily nested iterable.
 
     Args:
-        nested_list:
-            An arbitrarily nested iterable to be flattened.
+        nested_iterable:
+            The nested iterable to be flattened.
 
-    Returns:
-        A list containing a flattened version of the arbitrarily nested input.
+    Yields:
+        The flattened items from the nested iterable.
     """
-    flat_list = []
-    if not isinstance(nested_list, (list, tuple)):
-        raise ValueError(
-            f"Expected object of type list or tuple, not {type(nested_list)}"
-        )
-    for item in nested_list:
-        if isinstance(item, (list, tuple)):
-            flat_list.extend(flatten(item))
+    if not isinstance(nested_iterable, Iterable) or isinstance(
+        nested_iterable, (str, bytes)
+    ):
+        yield nested_iterable
+        return
+    for item in nested_iterable:
+        if isinstance(item, Iterable) and not isinstance(item, (str, bytes)):
+            yield from flatten(item)
         else:
-            flat_list.append(item)
-    return flat_list
+            yield item
```

Comment:
I was surprised that this fundamental change to the return type from a list to a generator had few knock on effects. The returned object can only be used once, it cannot be indexed, etc. without conversion to a list. I note that you've wrapped the output in a `list()` in most places, so that explains that.

The only comment I have is that the doc-string here:
https://github.com/metoppv/improver/blob/dee4f9b1ac20b6c3adfab0d124fdc5be54b6e344/improver/utilities/forecast_reference_enforcement.py#L333
ought to reflect that `cubes` could be a generator, given that the only place this actually gets called (normalise_to_reference) now passes in a generator returned by the `flatten` function.